catchup/service.go

```diff
@@ -456,15 +458,13 @@ func (s *Service) pipelinedFetch(seedLookback uint64) {
 
 	for {
 		for nextRound < firstRound+basics.Round(parallelRequests) {
-			if s.nextRoundIsNotSupported(nextRound) {
-				// We may get here when the service starts
-				// and gets to an unsupported round.  Since in
-				// this loop we do not wait for the requests
-				// to be written to the ledger, there is no
-				// guarantee that the unsupported round will be
-				// stopped in this case.
-				s.handleUnsupportedRound(nextRound)
-				return
+			if s.roundIsNotSupported(nextRound) {
+				// Break out of the loop to avoid fetching
+				// blocks that we don't support.  If there
+				// are no more supported blocks to fetch,
+				// s.unsupportedRoundMonitor() will cancel
```

Comment:
not sure if the monitor even needed - this outer loop can do
```golang
		nextRound := s.ledger.NextRound()
		if s.roundIsNotSupported(nextRound) {
			return / s.cancel()
		}
```
right after the switch statement - similar to the goroutine it will be called periodically, in this case on each fetch completion. Since fetches complete sequentially, it will trigger after the last supported block written.

Reply:
One benefit of the monitor is that it avoids having to do the same cancellation logic in another place (`s.fetchRound()`), and it also seems more obviously correct and deadlock-free.  This isn't a particularly strong argument; I did initially try to implement the cancellation logic inline inside that loop, but ended up with some kind of deadlock due to a race, where the ledger didn't yet have the latest block when checking in the loop, but then got blocked waiting in the select loop, which led me to just simplify it using a separate monitor goroutine.  Certainly fixable though, so if you have a strong preference to avoid the monitor goroutine, I can go back and make the inline version work too (and call it from the `s.fetchRound()` code that checks for unsupported blocks).

Reply:
I do not feel strongly, just a feeling we have too much goroutines and it might already penalizing us (still working on confirmation tho).

Reply:
okay, I confirmed my proposal does not really work, `TestOnSwitchToUnSupportedProtocol2` passes and `TestOnSwitchToUnSupportedProtocol4` fails. Approved.

```diff
diff --git a/catchup/service.go b/catchup/service.go
index 7870f0ded..c1a4e562f 100644
--- a/catchup/service.go
+++ b/catchup/service.go
@@ -463,8 +463,7 @@ func (s *Service) pipelinedFetch(seedLookback uint64) {
                                // to be written to the ledger, there is no
                                // guarantee that the unsupported round will be
                                // stopped in this case.
-                               s.handleUnsupportedRound(nextRound)
-                               return
+                               break
                        }
 
                        done := make(chan bool, 1)
@@ -502,6 +501,9 @@ func (s *Service) pipelinedFetch(seedLookback uint64) {
                case <-s.ctx.Done():
                        return
                }
+               if s.nextRoundIsNotSupported(s.ledger.NextRound()) {
+                       s.cancel()
+               }
        }
 }
```